### DISCUSSION TITLE
Logically "Or"ed OAuth Scopes

### DISCUSSION THREAD

**Discussion by jeffsawatzky** [fastapi/fastapi#11483] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import Depends, FastAPI, Security
from fastapi.security import OAuth2PasswordBearer, SecurityScopes

oauth2_scheme = OAuth2PasswordBearer(
    tokenUrl="token",
    scopes={"me": "Read information about the current user.", "items": "Read items."},
)

app = FastAPI()


class User:
    def __init__(self, username: str):
        self.username = username


async def get_current_user(
    security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]
):
    # TODO: Do something with the security scopes and token
    return User(username="Joe")


#
# How do I allow one ANY of the following scopes: `admin`, `item:read`, `all:read`?
#
@app.get("/users/me/items/")
async def read_own_items(
    current_user: Annotated[User, Security(get_current_user, scopes=["items"])],
):
    return [{"item_id": "Foo", "owner": current_user.username}]
```

### Description

If I have an endpoint that requires the OAuth token to have either the "admin" scope OR "item:read" scope OR "all:read" scope, how would I express this with the `Security` and `SecurityScopes` dependencies?

The OAS spec seems to allow this
https://swagger.io/docs/specification/authentication/#multiple

And it would look something like this
```
security:
  - oauth2: [admin]
  - oauth2: [item:read]
  - oauth2: [all:read]
```

https://github.com/OAI/OpenAPI-Specification/discussions/3001#discussioncomment-3456275

Thanks!

### Operating System

macOS

### FastAPI Version

0.110.1

### Pydantic Version

1.10.15

### Python Version

3.11

**Comment by YuriiMotov** [comment]
I tried to implement what you want, but i didn't succeed.
I think FastAPI doesn' support this and it's impossible in current FastAPI implementation.
It would be a good feature to implement in future FastAPI versions!

The only way I found is:
```
from typing import Annotated

from fastapi import Depends, FastAPI, HTTPException, Security
from fastapi.security import OAuth2PasswordBearer, SecurityScopes

oauth2_scheme = OAuth2PasswordBearer(
    tokenUrl="token",
    scopes={"admin": "", "item:read": "", "all:read": ""},
    auto_error=False
)

app = FastAPI()


class User:
    def __init__(self, username: str):
        self.username = username


async def get_current_user(
    security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]
):
    # TODO: Do something with the security scopes and token
    print(security_scopes.scope_str)
    return User(username="Joe")


#
# How do I allow one ANY of the following scopes: `admin`, `item:read`, `all:read`?
#
@app.get("/users/me/items/")
async def read_own_items(
    current_user_admin: Annotated[User, Security(get_current_user, scopes=["items"])],
    current_user_item_read: Annotated[User, Security(get_current_user, scopes=["item:read"])],
    current_user_all_read: Annotated[User, Security(get_current_user, scopes=["all:read"])],
):
    current_user = current_user_admin or current_user_all_read or current_user_item_read
    if current_user is None:
        raise HTTPException(status_code=403)
    return [{"item_id": "Foo", "owner": current_user.username}]
```

This produces openapi:
```
        "security": [
          {
            "OAuth2PasswordBearer": [
              "items"
            ]
          },
          {
            "OAuth2PasswordBearer": [
              "item:read"
            ]
          },
          {
            "OAuth2PasswordBearer": [
              "all:read"
            ]
          }
        ]
```

But I don't think it's usefull and somebody will use this approach..
